src/catalog/src/durable/upgrade.rs

```diff
@@ -297,4 +313,74 @@ pub(crate) mod persist {
 
         Ok(upper)
     }
+
+    /// Runs `migration_logic` on the contents of the current catalog assuming a current version of
+    /// `current_version`.
+    ///
+    /// Returns the new version and upper.
+    async fn run_versioned_upgrade<V1: IntoStateUpdateKindBinary, V2: IntoStateUpdateKindBinary>(
+        persist_handle: &mut PersistHandle,
+        upper: Timestamp,
+        current_version: u64,
+        migration_logic: impl FnOnce(Vec<V1>) -> Vec<MigrationAction<V1, V2>>,
+    ) -> Result<(u64, Timestamp), CatalogError> {
+        // 1. Get a snapshot of the current catalog, using the V1 to deserialize the
+        // contents.
+        let as_of = persist_handle.as_of(upper);
+        let snapshot = persist_handle.snapshot_binary(as_of).await;
+        let snapshot: Vec<_> = snapshot
+            .into_iter()
+            .map(|update| {
+                soft_assert_eq!(1, update.diff, "snapshot is consolidated");
+                V1::try_from(update.clone().kind).expect("invalid catalog data persisted")
+            })
+            .collect();
+
+        // 2. Generate updates from version specific migration logic.
+        let migration_actions = migration_logic(snapshot);
+        let mut updates: Vec<_> = migration_actions
+            .into_iter()
+            .flat_map(|action| action.into_updates(upper).into_iter())
+            .collect();
+
+        // 3. Add a retraction for old version and insertion for new version into updates.
+        let next_version = current_version + 1;
+        let version_retraction = StateUpdate {
+            kind: version_update_kind(current_version),
+            ts: upper,
+            diff: -1,
+        };
+        updates.push(version_retraction);
+        let version_insertion = StateUpdate {
+            kind: version_update_kind(next_version),
+            ts: upper,
+            diff: 1,
+        };
+        updates.push(version_insertion);
+
+        // 4. Compare and append migration into persist shard.
+        let next_upper = upper.step_forward();
+        persist_handle
+            .compare_and_append(updates, upper, next_upper)
+            .await?;
+
```

Comment:
I just realized that we need to force compaction here or else we may still try and deserialize retractions from old versions and fail.

Reply:
Actually we don't need to do this, because when we read, we always read a consolidated snapshot.